File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    long long calculateTotalHours(vector<int>& v, int hourly) {
        long long totalH = 0;

        for (int bananas : v) {
            totalH += (bananas + hourly - 1LL) / hourly;
        }

        return totalH;
    }

    int minEatingSpeed(vector<int>& v, int h) {
        int low = 1;
        int high = *max_element(v.begin(), v.end());

        while (low <= high) {
            int mid = low + (high - low) / 2;

            long long totalH = calculateTotalHours(v, mid);

            if (totalH <= h)
                high = mid - 1;
            else
                low = mid + 1;
        }

        return low;
    }
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        st=1
       
        end =max(piles)
        while st<=end:
            mid=st+(end-st)//2
            hours=0
            for pile in piles:
                hours +=(pile+mid-1)//mid
            if hours<=h:
                end=mid-1
            else:
                st=mid+1
        return st
            
        